#setting working directory
setwd("~/Documents/GitHub/scRNA-seq_VDJ/")

#install R packages, only need to run on inititation of the project
#install.packages("tidyverse")

#loading R packages
library(tidyverse)

#loading the post-processed IgBLAST datasets
## here loading from github repo, to load from file change from URL to file location
### PBMC, loading the Ig and TR data
pbmc.b.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/pbmcs/sc5p_v2_hs_PBMC_10k_b_filtered_contig.ig.igblast.tsv")
pbmc.t.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/pbmcs/sc5p_v2_hs_PBMC_10k_t_filtered_contig.tr.igblast.tsv")
### tumour, loading the Ig and TR data
tumour.b.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/tumour/sc5p_v2_hs_melanoma_10k_b_filtered_contig.ig.igblast.tsv")
tumour.t.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/tumour/sc5p_v2_hs_melanoma_10k_t_filtered_contig.tr.igblast.tsv")

#loading the cell ranger contig annotations
## here, loading from URL
### PBMCs
pbmc.b.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/5.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_b_filtered_contig_annotations.csv")
pbmc.t.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/5.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_filtered_contig_annotations.csv")

### tumour
tumour.b.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_melanoma_10k/sc5p_v2_hs_melanoma_10k_b_filtered_contig_annotations.csv")
tumour.t.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_melanoma_10k/sc5p_v2_hs_melanoma_10k_t_filtered_contig_annotations.csv")

#take a look at the data which is currently data for each contig, eg. AAACCTGTCATATCGG-1_contig_1 where AAACCTGTCATATCGG-1 is the cell barcode
pbmc.b.igblast %>% glimpse()
pbmc.b.cellranger %>% glimpse()
#for IgBLAST the contig label is in the sequence_id column, for cellranger it is in the contig_id column

#merging the cellranger and igblast
## as part of merging want to retain which columns were generated by IgBLAST and cellranger by appending either _cellranger or _igblast to the column names
## some columns are excluded from the rename_with, those in the -c() are excluded so not adding the _cellranger to contig_id & barcode from the cellranger tibble
## or for the sequence_id from the IgBLAST tibble
## the tibbles are then joined by the contig_id (cellranger) and sequence_id (igblast) columns
pbmc.b.data <- pbmc.b.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(pbmc.b.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

pbmc.t.data <- pbmc.t.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(pbmc.t.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

tumour.b.data <- tumour.b.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(tumour.b.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

tumour.t.data <- tumour.t.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(tumour.t.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

#can remove the original tibbles now that we have the merged objects
#rm(list = c("pbmc.b.cellranger", "pbmc.b.igblast", "pbmc.t.cellranger", "pbmc.t.igblast",
#            "tumour.b.cellranger", "tumour.b.igblast", "tumour.t.cellranger", "tumour.t.igblast"))

#can join all the Ig data into single tibble, adding a "dataset" column to track whether pbmc or tumour sample
b.data <- bind_rows(pbmc.b.data %>% mutate(dataset = "pbmc"),
                    tumour.b.data %>% mutate(dataset = "tumour"))

t.data <- bind_rows(pbmc.t.data %>% mutate(dataset = "pbmc"),
                    tumour.t.data %>% mutate(dataset = "tumour"))

#how many contigs per dataset
table(b.data$dataset)
table(t.data$dataset)

#remove the individual datasets
#rm(list = c("pbmc.b.data", "tumour.b.data",
#            "pbmc.t.data", "tumour.t.data"))

###################
###   TR data   ###
###################

#have TR contigs from TRA and TRB loci, want to start with some filtering of the contigs from each locus
#going to drop non-productive (out-of-frame, stop codons), those without both a V and J gene, without defined CDR3
#then for each cell barcode the contigs are ranked by their UMI counts from the cell ranger data
#also going to add a clonotype label that combines the V + J + CDR3AA

#TRA chains
tra.data <- t.data %>%
  filter(locus_igblast == "TRA") %>% #restrict to TRA chains
  filter(grepl("TRAV", v_call_igblast)) %>% #must have a TRAV gene
  filter(grepl("TRAJ", j_call_igblast)) %>% #must have a TRAJ gene
  filter(!(is.na(cdr3_aa_igblast))) %>% # check if the cdr3_aa column is NA and then keep if it is NOT (!) NA
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% #stop codons are encoded by *, * is a special character and needs to be escaped with \\
  filter(productive_igblast == TRUE) %>% #keep only sequences flagged as productive by IgBLAST
  filter(stop_codon_igblast == FALSE) %>% #keep only sequences lacking stop codons, note IgBLAST only flags this for V region, hence the line above
  filter(vj_in_frame_igblast == TRUE) %>% #keep only sequence that are in-frame for V-J, if J is out-of-frame, Ig is non-productive
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% #when designating clonotypes, don't use the allele, this creates versions of gene names that remove the allele
  mutate(clonotype = paste(v, j, cdr3_aa_igblast, sep = "_")) %>% #note that CDR3 from IMGT doesn't include anchors, may be preferable to use junction in some cases
  group_by(barcode, dataset) %>% #to do the ranking we want to group by barcode, as we combined two datasets barcodes may not be unique so adding dataset too
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% #rank for the contig by descending order of UMI counts from cell ranger, using row_number to avoid dups
  ungroup() #remove the grouping by barcode + dataset

table(tra.data$dataset)

#TRB chains
trb.data <- t.data %>%
  filter(locus_igblast == "TRB") %>% 
  filter(grepl("TRBV", v_call_igblast)) %>% 
  filter(grepl("TRBJ", j_call_igblast)) %>% 
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% 
  mutate(clonotype = paste(v, j, cdr3_aa_igblast, sep = "_")) %>% 
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

## for each cell going to join the TRA and TRB chain data (only keeping the top ranked contig from TRA and TRB per barcode)
tcr.data <- tra.data %>% 
  filter(contig_rnk_in_cln == 1) %>% #filtering the TRA data to only the top ranked contigs
  full_join(trb.data %>% filter(contig_rnk_in_cln == 1), #filter the TRB data for only top ranked, doing a full join as there may be barcodes in one and not the other
            suffix = c("_tra", "_trb"), #as cols will be duplicated, adding _tra and _trb
            by = c("barcode", "dataset")) #joining by the barcode & dataset as barcode alone may not be unique

#get a count for the number of cells with TRA/B data 
with(tcr.data, table(dataset))

###################
###   Ig data   ###
###################
# for the Ig there is IGH and then light chain can be either IGK or IGL, as want to pair heavy and light chains per cell, keep IGH and then pick a top chain from IGK/L after filtering
# applying similar filtering to the TR, but no clonotype label, instead will build clones at a later point as want to allow for SHM

#IGH chains
igh.data <- b.data %>%
  filter(locus_igblast == "IGH") %>% 
  filter(grepl("IGHV", v_call_igblast)) %>% 
  filter(grepl("IGHJ", j_call_igblast)) %>% 
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% #not designating clones at this time but these columns will be useful to have
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

table(igh.data$dataset)

#IGK/L chains
igkl.data <- b.data %>%
  filter(locus_igblast == "IGK" | locus_igblast == "IGL") %>% #keeping both IGK and IGL by allowing the locus columns to be IGK OR (|) IGL
  filter(grepl("IG[KL]V", v_call_igblast)) %>% #grepl is using a regular expression, by including the [KL] it allows for either of these letters at the 3rd position
  filter(grepl("IG[KL]J", j_call_igblast)) %>% #grepl is using a regular expression, by including the [KL] it allows for either of these letters at the 3rd position
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% 
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

table(igkl.data$dataset)

## for each cell going to join the IGH and IGK/L chain data (only keeping the top ranked contig from IGH and IGK/L per barcode)
bcr.data <- igh.data %>% 
  filter(contig_rnk_in_cln == 1) %>% #filtering the IGH data to only the top ranked contigs
  full_join(igkl.data %>% filter(contig_rnk_in_cln == 1), #filter the IGK/L data for only top ranked, doing a full join as there may be barcodes in one and not the other
            suffix = c("_igh", "_igkl"), #as cols will be duplicated, adding _igh and _igkl
            by = c("barcode", "dataset")) #joining by the barcode & dataset as barcode alone may not be unique

#get a count for the number of cells with Ig data 
with(bcr.data, table(dataset))

#saving the Ig data for the datasets
## writing tab-delimited files
write_tsv(bcr.data %>% filter(dataset == "pbmc"), "pbmc_Ig_cellranger_igblast_per-barcode.tsv")
write_tsv(bcr.data %>% filter(dataset == "tumour"), "tumour_Ig_cellranger_igblast_pre-barcode.tsv")
write_tsv(bcr.data, "pbmc-tumour_Ig_cellranger_igblast_per-barcode.tsv")

write_tsv(tcr.data %>% filter(dataset == "pbmc"), "pbmc_TR_cellranger_igblast_per-barcode.tsv")
write_tsv(tcr.data %>% filter(dataset == "tumour"), "tumour_TR_cellranger_igblast_pre-barcode.tsv")
write_tsv(tcr.data, "pbmc-tumour_TR_cellranger_igblast_per-barcode.tsv")

##writing R objects
saveRDS(bcr.data, "bcr_data.Rds")
saveRDS(tcr.data, "tcr_data.Rds")

