---
title: "DGE using VDJ features"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

Combing the metadata from the VDJ assay with the `FindMarkers()` function from `Seurat` permits differential gene expression (DGE) comparisons using the VDJ features such as expanded vs. unexpanded clones, comparing clones/clonotypes to each other, or comparing groups of cells with particular shared VDJ features like V isotype or germline gene usage.

The RScript for this workflow is available on [github](https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/RScripts/dge_using_VDJ_features.R) in the [workshop respository](https://github.com/kjlj/scRNA-seq_VDJ).

### Setting up R session

### R packages

For this need the Seurat package, the tidyverse packages, ggsci for colour pallettes and ggrepel for repelling labels from points on the 'volcano' plots.

```{r loading_r_packages}
#packages

#check if the package is already installed, if not, install it
if (sum(grepl("Seurat", rownames(installed.packages()))) > 0) {
  print("tidyverse is already installed")
} else {
  install.packages("Seurat")
}
#loading R packages
library(Seurat)

#check if the package is already installed, if not, install it
if (sum(grepl("tidyverse", rownames(installed.packages()))) > 0) {
  print("tidyverse is already installed")
} else {
  install.packages("tidyverse")
}
#loading R packages
library(tidyverse)

#check if the package is already installed, if not, install it
if (sum(grepl("ggsci", rownames(installed.packages()))) > 0) {
  print("ggsci is already installed")
} else {
  install.packages("ggsci")
}
#loading R packages
library(ggsci)

#check if the package is already installed, if not, install it
if (sum(grepl("ggrepel", rownames(installed.packages()))) > 0) {
  print("ggrepel is already installed")
} else {
  install.packages("ggrepel")
}
#loading R packages
library(ggrepel)
```

### Themes for plotting

Custom themes for ggplot2:
```{r plotting_themes}
#themes for plots
theme_custom <- function (base_size = 12) { 
  theme_bw(base_size=base_size) %+replace% 
    theme(panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_line(colour="black"),
          axis.text.x = element_text(colour="black", size=base_size),
          axis.text.y = element_text(colour="black", size = base_size, hjust = 1),
          axis.title = element_text(colour="black", face="bold", size = base_size),
          strip.background = element_blank(),
          strip.text = element_text(colour="black", face="bold", size=base_size),
          legend.text = element_text(colour="black", size=base_size)
    )
}

theme_x_rotated <- function (base_size = 12) { 
  theme_custom(base_size=base_size) %+replace% 
    theme(axis.text.x = element_text(colour="black", angle = 90, hjust = 1, vjust = 0.5, size=base_size))
}
```

## Loading datasets

Loading the seurat object plus the metadata:
```{r loading_dataset}
#loading seurat object
tumour.seurat.obj <- readRDS("../seurat_objects/tumour_seurat-without-VDJ-genes-azimuth.rds")

#loading the seurat exported data
seurat.data <- read_tsv("../data/pbmc-tumour_seruat_data.tsv")
```

Adding metadata to the seurat object:
```{r adding_metadata}
#add the metadata to seurat object
metadata <- as.data.frame(seurat.data %>% filter(dataset == "tumour"))
rownames(metadata) <- metadata$barcode
metadata <- metadata %>% select(-barcode)

tumour.seurat.obj <- AddMetaData(tumour.seurat.obj, metadata)
```

## Comparing gene markers between different groups of cells

Quick plot of some of the annotation on the UMAP:
```{r quick_umap}
#umap of T subsets cells by Azimuth annotation
azimuth_levels <- c(unique(seurat.data$azimuth_l2)[grepl("^CD[4|8]", unique(seurat.data$azimuth_l2))], "MAIT", "Treg")

#add to the data
seurat.data <- seurat.data %>% 
  mutate(cell_type = ifelse(azimuth_l2 %in% azimuth_levels, paste(azimuth_l2), "other"))

#plotting
p1 <- ggplot(seurat.data, 
       aes(x = umap_1, y = umap_2)) +
  geom_point(data = seurat.data %>% filter(vdj_cell_type != "T"), 
             size = 0.5, colour = "lightgray") +
  geom_point(data = seurat.data %>% filter(vdj_cell_type == "T"),
             size = 0.8,
             aes(colour = cell_type)) +
  scale_colour_d3(palette = "category20",
                  name = "cell type") +
  theme_custom() +
  guides(colour = guide_legend(override.aes = list(size = 5)))
print(p1)
```

The FindMarkers function from the Seurat package can be used for comparing gene expression between groups of cells. 

We have the T cell annotation from Azimuth, so can use this to do a basic comparison between cells from the naive and central memory populations within the CD4s. To do this the FindMarkers function is called with 'azimuth_l2' as the grouping and then ident.1 and ident.2 are used to set the labels for the azimuth_l2 that are to be compared, here ident.1 is set to "CD4 Naive" and ident.2 to the "CD4 TCM":
```{r markers_1}
cd4naive.vs.cd4tcm.markers <- FindMarkers(tumour.seurat.obj,  
                                          group.by = "azimuth_l2", 
                                          ident.1 = "CD4 Naive", ident.2 = "CD4 TCM") %>% 
  rownames_to_column("gene") %>%
  arrange(desc(avg_log2FC))

#filter for significance and print the top genes by logFC
cd4naive.vs.cd4tcm.markers %>% 
  filter(p_val_adj < 0.05) %>% 
  head()
```

In addition to the tabular format, can use the output to create volcano plots:
```{r volcano_1}
#show as a volcano plot
p1 <- ggplot(cd4naive.vs.cd4tcm.markers, 
       aes(x = avg_log2FC, y = -log10(p_val_adj))) +
  geom_point(aes(colour = ifelse(p_val_adj < 0.05, "sig", "ns"))) +
  geom_text_repel(aes(label = ifelse(p_val_adj < 0.05, gene, NA))) +
  theme_custom() +
  scale_colour_manual(values = c("firebrick", "darkgray"),
                      name = "",
                      limits = c("sig", "ns")) +
  guides(colour = guide_legend(override.aes = list(size = 5))) +
  labs(x = "average log2 fold change",
       y = "-log10(adjusted p-value)") 
plot(p1)  
```

An example of comparing the expanded versus singleton clones within the T cells. To do this we used the t_expanded_cln column as the group.by and ident.1 and ident.2 as 'expanded' and 'singleton', respectively.
```{r markers2}
#can compare expanded versus singleton clones
texp.vs.tsingle.markers <- FindMarkers(tumour.seurat.obj,  
                                          group.by = "t_expanded_cln", 
                                          ident.1 = "expanded", ident.2 = "singleton") %>% 
  rownames_to_column("gene") %>%
  arrange(desc(avg_log2FC))

#printing top genes
texp.vs.tsingle.markers %>% 
  filter(p_val_adj < 0.05) %>% 
  head()
```

The volcono plot for comparing gene markers between the expanded T cells and singleton T cells:
```{r volcano2}
#show as a volcano plot
p1 <- ggplot(texp.vs.tsingle.markers, 
       aes(x = avg_log2FC, y = -log10(p_val_adj))) +
  geom_point(aes(colour = ifelse(p_val_adj < 0.05, "sig", "ns"))) +
  geom_text_repel(aes(label = ifelse(p_val_adj < 0.05, gene, NA))) +
  theme_custom() +
  scale_colour_manual(values = c("firebrick", "darkgray"),
                      name = "",
                      limits = c("sig", "ns")) +
  guides(colour = guide_legend(override.aes = list(size = 5))) +
  labs(x = "average log2 fold change",
       y = "-log10(adjusted p-value)")
plot(p1)
```


Can take this to the level of comparing individual clones to other groups of cells. The largest T cell clone in the tumour is only 7 cells, but there are instances where you have may have very large expansions to compare. In this case, using the t_cln_rnk to compare the top clone to the second largest clone, but can use the paired_clontypes if had labels for particular clones of interest:
```{r markers3}
no1.vs.no2.markers <- FindMarkers(tumour.seurat.obj,  
                                       group.by = "t_cln_rnk", 
                                       ident.1 = "1", ident.2 = "2") %>% 
  rownames_to_column("gene") %>%
  arrange(desc(avg_log2FC))

no1.vs.no2.markers %>% head()

no1.vs.no2.markers %>% 
  filter(p_val_adj < 0.05) %>% 
  head()
```
No markers to differential the top 2 T cell clonotypes were found.

The comparison need not be limited to one vs one, can use vectors for ident.1 and ident.2 to create groups that span multiple categories. For example, sticking with t_cln_rnk but now grouping the top 2 clonotypes (ident.1 = c("1", "2")) and comparing those clonotypes ranked between 3 - 5 (ident.2 = c("3", "5")):
```{r markers4}
## this will compare the clones ranked 1 and 2 with those ranked 3 - 5
grp1.vs.grp2.markers <- FindMarkers(tumour.seurat.obj,  
                                  group.by = "t_cln_rnk", 
                                  ident.1 = c("1", "2"), ident.2 = c("3", "5")) %>% 
  rownames_to_column("gene") %>%
  arrange(desc(avg_log2FC))

grp1.vs.grp2.markers %>% head()

grp1.vs.grp2.markers %>% 
  filter(p_val_adj < 0.05) %>% 
  head()
```
In this case, there were not genes that differentiate the clonotypes groups.

Can use any of the metadata columns, for example, can compare the gene expression between B cells that utilise different VH genes in their IGHs. To do so, v_igh becomes the group.by and two IGHV genes are specified for ident.1 and ident.2. In this case, comparing IGHV1-69 expressing B cells to those utilising IGHV3-23:
```{r markers5}
vgene1.vs.vgene2.markers <- FindMarkers(tumour.seurat.obj,  
                                    group.by = "v_igh", 
                                    ident.1 = c("IGHV1-69"), ident.2 = c("IGHV3-23")) %>% 
  rownames_to_column("gene") %>%
  arrange(desc(avg_log2FC))

vgene1.vs.vgene2.markers %>% head()

vgene1.vs.vgene2.markers %>% 
  filter(p_val_adj < 0.05) %>% 
  head()
```

View on a volcano style plot:
```{r volcano5}
p1 <- ggplot(vgene1.vs.vgene2.markers, 
             aes(x = avg_log2FC, y = -log10(p_val_adj))) +
  geom_point(aes(colour = ifelse(p_val_adj < 0.05, "sig", "ns"))) +
  geom_text_repel(aes(label = ifelse(p_val_adj < 0.05, gene, NA))) +
  theme_custom() +
  scale_colour_manual(values = c("firebrick", "darkgray"),
                      name = "",
                      limits = c("sig", "ns")) +
  guides(colour = guide_legend(override.aes = list(size = 5))) +
  labs(x = "average log2 fold change",
       y = "-log10(adjusted p-value)") 
plot(p1)  
```

Based on the comparison IgM and FCER2 more common among IGHV1-69 utilising B cells which may suggest that this gene is more commonly utilised in naive cells compared to IGHV3-23.

[Return to main page](https://kjlj.github.io/scRNA-seq_VDJ/)
