---
title: "Joining Cell Ranger and IgBLAST VDJ outputs"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview

For each dataset now have the V(D)J data from Cell Ranger (the filtered_contig_annotations.csv files) and IgBLAST (the igblast.tsv files). 

Combining these files in R to summarise the paired chains for each cell. Running R in RStudio on [Posit cloud](https://posit.cloud/). After logging in, can create a new project and within the project a new RScript (joining_cellranger_igblast.R).

Workflow for combining files is described [here](https://kjlj.github.io/scRNA-seq_VDJ/docs/joining_cellranger_igblast.html). The RScript is available on [github](https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/RScripts/joining_cellranger_igblast.R) in the [workshop respository](https://github.com/kjlj/scRNA-seq_VDJ).

## Working locally versus working on Posit Cloud

If working locally it may be of benefit to set the working directory at the start of your R session so that you can use relative file paths, this isn't required when using RStudio on Posit Cloud.

Uncomment set working directory function and update the path to the location of your choice:
```{r set_working_dir}
#setwd("~/Documents/GitHub/scRNA-seq_VDJ/")
```

## R packages

The [tidyverse collection](https://www.tidyverse.org/) of R packages will be used to manipulate the datasets. To install this package (if it is not already available) and then load:
```{r install_packages}
#check if the package is already installed, if not, install it
if (sum(grepl("tidyverse", rownames(installed.packages()))) > 0) {
  print("tidyverse is already installed")
} else {
  install.packages("tidyverse")
}

#loading R packages
library(tidyverse)
```

## Loading datasets

Loading the post-processed IgBLAST datasets from github repo for the B and T cell VDJ data for the PBMC  and Tumour samples. The Cell Ranger contig annotations are loaded from the 10x website. If these files have already been downloaded from github/10x the URLs can be replaced with the local file locations.

```{r loading_datasets}

### PBMC, loading the Ig and TR data
pbmc.b.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/data/pbmcs/sc5p_v2_hs_PBMC_10k_b_filtered_contig.ig.igblast.tsv")
pbmc.t.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/data/pbmcs/sc5p_v2_hs_PBMC_10k_t_filtered_contig.tr.igblast.tsv")
### tumour, loading the Ig and TR data
tumour.b.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/data/tumour/sc5p_v2_hs_melanoma_10k_b_filtered_contig.ig.igblast.tsv")
tumour.t.igblast <- read_tsv("https://raw.githubusercontent.com/kjlj/scRNA-seq_VDJ/main/data/tumour/sc5p_v2_hs_melanoma_10k_t_filtered_contig.tr.igblast.tsv")

#loading the cell ranger contig annotations
### PBMCs
pbmc.b.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/5.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_b_filtered_contig_annotations.csv")
pbmc.t.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/5.0.0/sc5p_v2_hs_PBMC_10k/sc5p_v2_hs_PBMC_10k_t_filtered_contig_annotations.csv")

### tumour
tumour.b.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_melanoma_10k/sc5p_v2_hs_melanoma_10k_b_filtered_contig_annotations.csv")
tumour.t.cellranger <- read_csv("https://cf.10xgenomics.com/samples/cell-vdj/4.0.0/sc5p_v2_hs_melanoma_10k/sc5p_v2_hs_melanoma_10k_t_filtered_contig_annotations.csv")
```

Take a look at the data that has been loaded where each entry in the tibbles are for a contig. For example, a single contig from a cell is AAACCTGTCATATCGG-1_contig_1 where AAACCTGTCATATCGG-1 is the cell barcode. Note that for IgBLAST the contig label is in the sequence_id column, for cellranger it is in the contig_id column.

The IgBLAST data for the PBMC dataset, for glimpse the data is printed as the first few entries from each column on each line so you can get a 'glimpse' at the values in each column:
```{r view_loaded_datasets}
pbmc.b.igblast %>% glimpse()
```

The Cell Ranger data for the PMBC dataset:
```{r view_loaded_datasets2}
pbmc.b.cellranger %>% glimpse()
```

Alternative is to view in a tabular format, but cannot see all the columns:
```{r view_loaded_datasets3}
pbmc.b.igblast %>% head()
```

## Merging the Cell Ranger and IgBLAST

The data for each contig can be joined using the contig_id in the sequence_id column for the IgBLAST and the contig_id column for Cell Ranger. 

As part of merging want to retain which columns were generated by IgBLAST and which were from Cell Ranger. This will be tracked by appending either _cellranger or _igblast to the column names. Some columns are excluded from the rename_with so those in the -c() will not have the suffix added. Not adding the _cellranger to contig_id & barcode from the cellranger tibble or for the sequence_id from the IgBLAST tibble:
```{r merge_datasets}
## the tibbles are then joined by the contig_id (cellranger) and sequence_id (igblast) columns
pbmc.b.data <- pbmc.b.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(pbmc.b.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

pbmc.t.data <- pbmc.t.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(pbmc.t.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

tumour.b.data <- tumour.b.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(tumour.b.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))

tumour.t.data <- tumour.t.cellranger %>% rename_with(~paste0(., "_cellranger"), -c("contig_id", "barcode")) %>%
  left_join(tumour.t.igblast %>% rename_with(~paste0(., "_igblast"), -c("sequence_id")),
            by = c("contig_id" = "sequence_id"))
```

Here is an example of the PBMC Ig data after joining the cellranger and IgBLAST:
```{r view_merged_dataset}
pbmc.b.data %>% glimpse()
```
the other datasets can be viewed using the glimpse(), head() or View() functions.


Optionally, can remove the original tibbles now that we have the merged objects. This can help save some memory.
```{r removing_orig_inputs}
rm(list = c("pbmc.b.cellranger", "pbmc.b.igblast", "pbmc.t.cellranger", "pbmc.t.igblast",
            "tumour.b.cellranger", "tumour.b.igblast", "tumour.t.cellranger", "tumour.t.igblast"))
```

We can reduce from four to two datasets to process by joining the B cell data from the PBMCs and tumour sample to one datasets and the same for the T cells. A dataset column is added to keep track of which contigs are from which datasets as we cannot rely on the cell barcodes being unique between the two datasets:
```{r joining_datasets}
#can join all the Ig data into single tibble, adding a "dataset" column to track whether pbmc or tumour sample
b.data <- bind_rows(pbmc.b.data %>% mutate(dataset = "pbmc"),
                    tumour.b.data %>% mutate(dataset = "tumour"))

t.data <- bind_rows(pbmc.t.data %>% mutate(dataset = "pbmc"),
                    tumour.t.data %>% mutate(dataset = "tumour"))
```

Check how many contigs are in each dataset for Ig:
```{r joined_b_tbl}
table(b.data$dataset)
```

And for the TR contigs:
```{r joined_t_tbl}
table(t.data$dataset)
```

Taking a peek at the contents of the TR data after combining the PBMC and Tumour datasets:
```{r view_joined_tr_data}
t.data %>% glimpse()
```

Optionally, can now drop the non-joined B and T cell datasets:
```{r removing_unjoined_tibbles}
#remove the individual datasets
rm(list = c("pbmc.b.data", "tumour.b.data",
            "pbmc.t.data", "tumour.t.data"))
```

## Processing the TR contigs
There are TR contigs from TRA and TRB loci from each cell (ideally!). Start with some filtering of the contigs from each locus in order to drop non-productive (out-of-frame, stop codons), those without both a V and J gene, without defined CDR3. Then for each cell barcode the contigs are ranked by their UMI counts from the cell ranger data and finally a clonotype label that combines the V + J + CDR3AA is added for each contig.

Starting with the TRA contigs that are selected based on the locus_igblast column:
```{r tra_filtering}
tra.data <- t.data %>%
  filter(locus_igblast == "TRA") %>% #restrict to TRA chains
  filter(grepl("TRAV", v_call_igblast)) %>% #must have a TRAV gene
  filter(grepl("TRAJ", j_call_igblast)) %>% #must have a TRAJ gene
  filter(!(is.na(cdr3_aa_igblast))) %>% # check if the cdr3_aa column is NA and then keep if it is NOT (!) NA
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% #stop codons are encoded by *, * is a special character and needs to be escaped with \\
  filter(productive_igblast == TRUE) %>% #keep only sequences flagged as productive by IgBLAST
  filter(stop_codon_igblast == FALSE) %>% #keep only sequences lacking stop codons, note IgBLAST only flags this for V region, hence the line above
  filter(vj_in_frame_igblast == TRUE) %>% #keep only sequence that are in-frame for V-J, if J is out-of-frame, Ig is non-productive
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% #when designating clonotypes, don't use the allele, this creates versions of gene names that remove the allele
  mutate(clonotype = paste(v, j, cdr3_aa_igblast, sep = "_")) %>% #note that CDR3 from IMGT doesn't include anchors, may be preferable to use junction in some cases
  group_by(barcode, dataset) %>% #to do the ranking we want to group by barcode, as we combined two datasets barcodes may not be unique so adding dataset too
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% #rank for the contig by descending order of UMI counts from cell ranger, using row_number to avoid dups
  ungroup() #remove the grouping by barcode + dataset

#how many contigs survive the filtering for each dataset (pbmc or tumour)
table(tra.data$dataset)
```

For the TRB contigs:
```{r trb_filtering}
#TRB chains
trb.data <- t.data %>%
  filter(locus_igblast == "TRB") %>% 
  filter(grepl("TRBV", v_call_igblast)) %>% 
  filter(grepl("TRBJ", j_call_igblast)) %>% 
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% 
  mutate(clonotype = paste(v, j, cdr3_aa_igblast, sep = "_")) %>% 
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

#how many contigs survive the filtering for each dataset (pbmc or tumour)
table(trb.data$dataset)
```

Finally for each cell join the TRA and TRB chain data, to get a single TRA and TRB chain for each cell barcode only keep the top ranked contig from TRA and TRB per barcode:
```{r join_tra_trb}
tcr.data <- tra.data %>% 
  filter(contig_rnk_in_cln == 1) %>% #filtering the TRA data to only the top ranked contigs
  full_join(trb.data %>% filter(contig_rnk_in_cln == 1), #filter the TRB data for only top ranked, doing a full join as there may be barcodes in one and not the other
            suffix = c("_tra", "_trb"), #as cols will be duplicated, adding _tra and _trb
            by = c("barcode", "dataset")) #joining by the barcode & dataset as barcode alone may not be unique

#get a count for the number of cells with TRA/B data 
with(tcr.data, table(dataset))

```

The tibble now contains TRA and TRB chains for each cell barcode with data from both the cellranger and igblast processing:
```{r view_tr_data}
tcr.data %>% glimpse()
```

## Processing the Ig contigs

For the Ig there is IGH and then light chain can be either IGK or IGL, as want to pair heavy and light chains per cell, keep IGH and then pick a top chain from IGK/L after filtering applying similar filtering to the TR, but no clonotype label, instead will build clones at a later point as want to allow for SHM. The clone building is described <insert link here>.

Processing the IGH contigs to filter and rank contigs from each cell barcode:
```{r igh_filtering}
#IGH chains
igh.data <- b.data %>%
  filter(locus_igblast == "IGH") %>% 
  filter(grepl("IGHV", v_call_igblast)) %>% 
  filter(grepl("IGHJ", j_call_igblast)) %>% 
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% #not designating clones at this time but these columns will be useful to have
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

#print a count of igh contigs to each dataset
table(igh.data$dataset)
```


For the light chains will use the data from both the IGK and IGL:
```{r igkl_filtering}
igkl.data <- b.data %>%
  filter(locus_igblast == "IGK" | locus_igblast == "IGL") %>% #keeping both IGK and IGL by allowing the locus columns to be IGK OR (|) IGL
  filter(grepl("IG[KL]V", v_call_igblast)) %>% #grepl is using a regular expression, by including the [KL] it allows for either of these letters at the 3rd position
  filter(grepl("IG[KL]J", j_call_igblast)) %>% #grepl is using a regular expression, by including the [KL] it allows for either of these letters at the 3rd position
  filter(!(is.na(cdr3_aa_igblast))) %>% 
  filter(!(grepl("\\*", sequence_alignment_aa_igblast))) %>% 
  filter(productive_igblast == TRUE) %>% 
  filter(stop_codon_igblast == FALSE) %>% 
  filter(vj_in_frame_igblast == TRUE) %>% 
  mutate(v = str_replace(v_call_igblast, "\\*.+$", ""),
         j = str_replace(j_call_igblast, "\\*.+$", "")) %>% 
  group_by(barcode, dataset) %>% 
  mutate(contig_rnk_in_cln = row_number(desc(umis_cellranger))) %>% 
  ungroup() 

table(igkl.data$dataset)
```


For each cell going to join the IGH and IGK/L chain data (only keeping the top ranked contig from IGH and IGK/L per barcode):
```{r joining_igh_igkl}
bcr.data <- igh.data %>% 
  filter(contig_rnk_in_cln == 1) %>% #filtering the IGH data to only the top ranked contigs
  full_join(igkl.data %>% filter(contig_rnk_in_cln == 1), #filter the IGK/L data for only top ranked, doing a full join as there may be barcodes in one and not the other
            suffix = c("_igh", "_igkl"), #as cols will be duplicated, adding _igh and _igkl
            by = c("barcode", "dataset")) #joining by the barcode & dataset as barcode alone may not be unique

#get a count for the number of cells with Ig data 
with(bcr.data, table(dataset))
```

The tibble now contains IGH and IGKL chains for each cell barcode with data from both the cellranger and igblast processing:
```{r view_ig_data}
bcr.data %>% glimpse()
```


## Outputting datasets

Saving the datasets in tab-delimited and RDS formats:
```{r output_datasets}
## writing tab-delimited files
write_tsv(bcr.data %>% filter(dataset == "pbmc"), "~/Documents/GitHub/scRNA-seq_VDJ/data/pbmcs/pbmc_Ig_cellranger_igblast_per-barcode.tsv")
write_tsv(bcr.data %>% filter(dataset == "tumour"), "~/Documents/GitHub/scRNA-seq_VDJ/data/tumour/tumour_Ig_cellranger_igblast_pre-barcode.tsv")
write_tsv(bcr.data, "~/Documents/GitHub/scRNA-seq_VDJ/data/pbmc-tumour_Ig_cellranger_igblast_per-barcode.tsv")

write_tsv(tcr.data %>% filter(dataset == "pbmc"), "~/Documents/GitHub/scRNA-seq_VDJ/data/pbmcs/pbmc_TR_cellranger_igblast_per-barcode.tsv")
write_tsv(tcr.data %>% filter(dataset == "tumour"), "~/Documents/GitHub/scRNA-seq_VDJ/data/tumour/tumour_TR_cellranger_igblast_pre-barcode.tsv")
write_tsv(tcr.data, "~/Documents/GitHub/scRNA-seq_VDJ/data/pbmc-tumour_TR_cellranger_igblast_per-barcode.tsv")

##writing R objects
saveRDS(bcr.data, "~/Documents/GitHub/scRNA-seq_VDJ/data/bcr_data.Rds")
saveRDS(tcr.data, "~/Documents/GitHub/scRNA-seq_VDJ/data/tcr_data.Rds")
```

## Next step

Now move to [building clones for the Ig data](https://kjlj.github.io/scRNA-seq_VDJ/docs/building_b_cell_clones.html).

Or [return to main page](https://kjlj.github.io/scRNA-seq_VDJ/)